README.markdown
# jQuery Chrony - A Count Down Plugin - http://wbotelhos.com/chrony

jQuery Chrony is a plugin that creates a chronometer.

## License

The jQuery Chrony is licensed under [The MIT License](http://www.opensource.org/licenses/mit-license.php)

## Version

	@version        0.2.0
	@since          2011.10.23
	@author         Washington Botelho
	@documentation  wbotelhos.com/chrony
	@twitter        twitter.com/wbotelhos

## Required Files

+ jquery.chrony.min.js

## Default values

	alert          : { second: 10 } // Attributes color, hour, minute and second to configure the alert.
	blink          : false          // Enables the colon blink.
	blinkTime      : 130            // Time spent to blink the colon.
	finish         : undefined      // Callback function to be executed when the time ends.
	decrement      : 1              // Number of seconds used to decrement the time.
	displayHours   : true           // Display or not the hours on clock. 
	displayMinutes : true           // Display or not the minutes on clock.
	displaySeconds : true           // Display or not the seconds on clock.
	hour           : 0              // Number of hours initial.
	hours          : undefined      // Number of hours initial auto adjusted to max if needed.
	minute         : 0              // Number of minutes initial.
	minutes        : undefined      // Number of minutes initial auto adjusted to hour if needed.
	second         : 0              // Number of seconds initial.
	seconds        : undefined      // Number of seconds initial auto adjusted to minute if needed.
	text           : undefined      // Time written out in full using the following mask: HH:mm:ss.

## Usage with default values

	$('#time').chrony({ hour: 1, minute: 2, second: 3 });

	<div id="time"></div>

## Buy me a coffee

You can do it by [PayPal](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=X8HEP2878NDEG&item_name=jQuery%20Chrony). Thanks! (:

## Contributors

Daniel Faria
Léo Hackin


changelog.markdown
# jQuery Chrony - A Count Down Plugin - http'//wbotelhos.com/chrony

### 0.2.0

+ Added option 'displayHours' to choose if the hours will be displayes; (by leohackin) 
+ Added option 'displayMinutes' to choose if the minutes will be displayes; (by leohackin)
+ Added option 'displaySeconds' to choose if the seconds will be displayes; (by leohackin)
+ Added method 'set' to re-adjust runtime options;
+ Now Chrony keeps the state of the elements and no longer depends of an ID.

### 0.1.0

+ Option 'alert' with color, hour, minute and second to configure the alert;
+ Option 'blink' to enables the colon blink;
+ Option 'blinkTime' to adjust the time spent to blink the colon;
+ Option 'finish' using as callback function to be executed when the time ends;
+ Option 'decrement' to change the number of seconds used to decrement the time;
+ Option 'hour' to change the number of hours initial;
+ Option 'hours' to change the number of hours initial auto adjusted to max if needed;
+ Option 'minute' to change the number of minutes initial;
+ Option 'minutes' to change the number of minutes initial auto adjusted to hour if needed;
+ Option 'second' to change the number of seconds initial;
+ Option 'seconds' to change the number of seconds initial auto adjusted to minute if needed;
+ Option 'text' to set the time written out in full using the following mask: HH:mm:ss.


test/spec.js
describe('Chrony', function() {

	beforeEach(function() {
	    $('body').append('<div id="time"></div>');
	});

	afterEach(function () {
		$('#time').remove();
	});

	describe('checking initialization', function() {

		it ('should mount the right clock', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ hour: 1, minute: 2, second: 3 });

			var	$firstColon		= $time.children('span').eq(0),
				$secondColon	= $time.children('span').eq(1);

			// then
		    expect($time).toContain('div#hour');
		    expect($time).toContain('div#minute');
		    expect($time).toContain('div#second');
		    expect($firstColon).toExist();
		    expect($secondColon).toExist();
		});

		it ('should initialize correctly with zeros', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ hour: 1, minute: 2, second: 3 });

			var $hour			= $time.children('#hour'),
				$firstColon		= $time.children('span').eq(0),
				$minute			= $time.children('#minute'),
				$secondColon	= $time.children('span').eq(1),
				$second			= $time.children('#second');

			// then
		    expect($hour).toHaveText('01');
		    expect($firstColon).toHaveText(':');
		    expect($minute).toHaveText('02');
		    expect($secondColon).toHaveText(':');
		    expect($second).toHaveText('03');
		});

		it ('should initialize correctly without zeros', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ hour: 10, minute: 20, second: 30 });

			var $hour			= $time.children('#hour'),
				$firstColon		= $time.children('span').eq(0),
				$minute			= $time.children('#minute'),
				$secondColon	= $time.children('span').eq(1),
				$second			= $time.children('#second');

			// then
		    expect($hour).toHaveText('10');
		    expect($firstColon).toHaveText(':');
		    expect($minute).toHaveText('20');
		    expect($secondColon).toHaveText(':');
		    expect($second).toHaveText('30');
		});

	});

	describe('auto adjusting the time', function() {

		it ('should adjust the seconds with hour minute and second', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ seconds: (1 * 60 * 60) + (2 * 60) + 3 });

			var $hour		= $time.children('#hour'),
				$minute		= $time.children('#minute'),
				$second		= $time.children('#second');

			// then
		    expect($hour).toHaveText('01');
		    expect($minute).toHaveText('02');
		    expect($second).toHaveText('03');
		});

		it ('should adjust the minutes with hour and minute', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ minutes:  (1 * 60) + 2 });

			var $hour		= $time.children('#hour'),
				$minute		= $time.children('#minute'),
				$second		= $time.children('#second');

			// then
		    expect($hour).toHaveText('01');
		    expect($minute).toHaveText('02');
		    expect($second).toHaveText('00');
		});

		it ('should adjust the hours with hour', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ hours:  23 });

			var $hour		= $time.children('#hour'),
				$minute		= $time.children('#minute'),
				$second		= $time.children('#second');

			// then
		    expect($hour).toHaveText('23');
		    expect($minute).toHaveText('00');
		    expect($second).toHaveText('00');
		});

		it ('should adjust the hours with max time', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ hours:  27 });

			var $hour		= $time.children('#hour'),
				$minute		= $time.children('#minute'),
				$second		= $time.children('#second');

			// then
		    expect($hour).toHaveText('23');
		    expect($minute).toHaveText('59');
		    expect($second).toHaveText('59');
		});

	});

	describe('checking the clock', function() {

		beforeEach(function() {
			jasmine.Clock.useMock();
		});

		it ('should mount the right clock', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ hour: 1, minute: 1, second: 1 });

			var $hour			= $time.children('#hour'),
				$firstColon		= $time.children('span').eq(0),
				$minute			= $time.children('#minute'),
				$secondColon	= $time.children('span').eq(1),
				$second			= $time.children('#second');

			// then
			expect($hour).toHaveText('01');
			expect($firstColon).toHaveText(':');
			expect($minute).toHaveText('01');
			expect($secondColon).toHaveText(':');
			expect($second).toHaveText('01');

			jasmine.Clock.tick(1000);
			expect($hour).toHaveText('01');
			expect($firstColon).toHaveText(':');
			expect($minute).toHaveText('01');
			expect($secondColon).toHaveText(':');
			expect($second).toHaveText('00');

			jasmine.Clock.tick(1000);
			expect($hour).toHaveText('01');
			expect($firstColon).toHaveText(':');
			expect($minute).toHaveText('00');
			expect($secondColon).toHaveText(':');
			expect($second).toHaveText('59');

			jasmine.Clock.tick(1000 * 59);
			expect($hour).toHaveText('01');
			expect($firstColon).toHaveText(':');
			expect($minute).toHaveText('00');
			expect($secondColon).toHaveText(':');
			expect($second).toHaveText('00');

			jasmine.Clock.tick(1000 * 60);
			expect($hour).toHaveText('00');
			expect($firstColon).toHaveText(':');
			expect($minute).toHaveText('59');
			expect($secondColon).toHaveText(':');
			expect($second).toHaveText('00');

			jasmine.Clock.tick(1000 * 60 * 59);
			expect($hour).toHaveText('00');
			expect($firstColon).toHaveText(':');
			expect($minute).toHaveText('00');
			expect($secondColon).toHaveText(':');
			expect($second).toHaveText('00');

			jasmine.Clock.tick(1000);
			expect($hour).toHaveText('00');
			expect($firstColon).toHaveText(':');
			expect($minute).toHaveText('00');
			expect($secondColon).toHaveText(':');
			expect($second).toHaveText('00');
		});

		it ('should call the callback', function() {
			// given
			var $time		= $('#time'),
				expected	= 'Finished!';

			// when
			$time.chrony({ seconds: 1, finish: function() { $(this).html(expected); } });

			// then
			jasmine.Clock.tick(1000);
			expect($time).toHaveText(expected);
		});

		it ('should call clock alert', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ seconds: 2, alert: { color: '#F00', hour: 0, minute: 0, second: 1 } });

			// then
			expect($time).not.toHaveAttr('style', 'color: rgb(255, 0, 0);');
			jasmine.Clock.tick(999);
			expect($time).not.toHaveAttr('style', 'color: rgb(255, 0, 0);');

			expect($time).not.toHaveAttr('style', 'color: rgb(255, 0, 0);');
			jasmine.Clock.tick(1);
			expect($time).toHaveAttr('style', 'color: rgb(255, 0, 0);');
		});

		it ('should call clock alert without color', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ second: 2, alert: { second: 1 } });

			// then
			expect($time).not.toHaveAttr('style', 'color: rgb(255, 0, 0);');
			jasmine.Clock.tick(1000);
			expect($time).toHaveAttr('style', 'color: rgb(255, 0, 0);');
		});

		it ('should call clock alert just with second', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ seconds: 2, alert: { second: 1 } });

			// then
			expect($time).not.toHaveAttr('style', 'color: rgb(255, 0, 0);');
			jasmine.Clock.tick(1000);
			expect($time).toHaveAttr('style', 'color: rgb(255, 0, 0);');
		});

		it ('should call clock alert just with minute', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ minute: 2, alert: { minute: 1 } });

			// then
			expect($time).not.toHaveAttr('style', 'color: rgb(255, 0, 0);');
			jasmine.Clock.tick(1000 * 60);
			expect($time).toHaveAttr('style', 'color: rgb(255, 0, 0);');
		});

		it ('should call clock alert just with hour', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ hour: 2, alert: { hour: 1 } });

			// then
			expect($time).not.toHaveAttr('style', 'color: rgb(255, 0, 0);');
			jasmine.Clock.tick(1000 * 60 * 60);
			expect($time).toHaveAttr('style', 'color: rgb(255, 0, 0);');
		});

		it ('should call clock alert with custom color', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ hour: 1, alert: { color: '#0F0' } });

			// then
			expect($time).not.toHaveAttr('style', 'color: rgb(0, 255, 0);');
			jasmine.Clock.tick(1000 * 60 * 60);
			expect($time).toHaveAttr('style', 'color: rgb(0, 255, 0);');
		});
		
		it ('should not display hours digits', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ minutes: 20, displayHours: false});

			// then
			var $firstColon		= $time.children('span').eq(0),
				$secondColon	= $time.children('span').eq(1);

			expect($time).not.toContain('div#hour');
			expect($time).toContain('div#minute');
			expect($time).toContain('div#second');
			expect($firstColon).toExist();
			expect($secondColon).not.toExist();
		});
		
		it ('should not display minutes digits', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ seconds: 20, displayMinutes: false});

			// then
			var $firstColon		= $time.children('span').eq(0),
				$secondColon	= $time.children('span').eq(1);

			expect($time).toContain('div#hour');
			expect($time).not.toContain('div#minute');
			expect($time).toContain('div#second');
			expect($firstColon).toExist();
			expect($secondColon).not.toExist();
		});
		
		it ('should not display seconds digits', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ minutes: 20, displaySeconds: false});

			// then
			var $firstColon		= $time.children('span').eq(0),
				$secondColon	= $time.children('span').eq(1);

			expect($time).toContain('div#hour');
			expect($time).toContain('div#minute');
			expect($time).not.toContain('div#second');
			expect($firstColon).toExist();
			expect($secondColon).not.toExist();
		});
		
		it ('should not display minutes and seconds digits', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ hours: 20, displayMinutes: false, displaySeconds: false});

			// then
			var $firstColon		= $time.children('span').eq(0),
				$secondColon	= $time.children('span').eq(1);

			expect($time).toContain('div#hour');
			expect($time).not.toContain('div#minute');
			expect($time).not.toContain('div#second');
			expect($firstColon).not.toExist();
			expect($secondColon).not.toExist();
		});

	});

	describe('public functions', function() {

		beforeEach(function() {
			jasmine.Clock.useMock();
		});

		it ('should override the finish callback', function() {
			// given
			var $time			= $('#time'),
				firstExpected	= 'Finished!',
				expected		= 'Overrided!';

			// when
			$time.chrony({ seconds: 1, finish: function() { $(this).html(firstExpected); } });

			$('#time').chrony('set', { finish: function() { $(this).html(expected); } });

			jasmine.Clock.tick(1000);

			// then
			expect($time).toHaveText(expected);
		});

		it ('should override decrement', function() {
			// given
			var $time = $('#time');

			// when
			$time.chrony({ seconds: 20 });

			$('#time').chrony('set', { decrement: 10 });

			// then
			var $hour			= $time.children('#hour'),
				$firstColon		= $time.children('span').eq(0),
				$minute			= $time.children('#minute'),
				$secondColon	= $time.children('span').eq(1),
				$second			= $time.children('#second');

			expect($hour).toHaveText('00');
			expect($firstColon).toHaveText(':');
			expect($minute).toHaveText('00');
			expect($secondColon).toHaveText(':');
			expect($second).toHaveText('20');

			jasmine.Clock.tick(1000);
			expect($hour).toHaveText('00');
			expect($firstColon).toHaveText(':');
			expect($minute).toHaveText('00');
			expect($secondColon).toHaveText(':');
			expect($second).toHaveText('10');

			jasmine.Clock.tick(1000);
			expect($hour).toHaveText('00');
			expect($firstColon).toHaveText(':');
			expect($minute).toHaveText('00');
			expect($secondColon).toHaveText(':');
			expect($second).toHaveText('00');
		});
	
	});

});


js/jquery.chrony.js
/*!
 * jQuery Chrony - A Count Down Plugin - http://wbotelhos.com/chrony
 * ------------------------------------------------------------------
 *
 * jQuery Chrony is a plugin that creates a chronometer.
 *
 * Licensed under The MIT License
 *
 * @version        0.2.0
 * @since          2011.10.23
 * @author         Washington Botelho
 * @documentation  wbotelhos.com/chrony
 * @twitter        twitter.com/wbotelhos
 *
 * Usage with default values:
 * ------------------------------------------------------------------
 * $('#time').chrony({ hour: 1, minute: 2, second: 3 });
 *
 * <div id="time"></div>
 *
 */

;(function($) {

	var methods = {
		init: function(options) {
			return this.each(function() {

				var self	= this,
					$this	= $(self);

				self.opt = $.extend(true, {}, $.fn.chrony.defaults, options);

				if ($this.data('chrony')) {
					return;
				}

				$this.data('chrony', true);

				var opt			= self.opt,
					separator	= '<span style="float: left;">:</span>';

				if (opt.text) {
					var text = opt.text.split(':');

					if (text.length != 3) {
						$.error('The format must be the following HH:mm:ss!');
					}

					opt.second = text[2];
					opt.minute = text[1];
					opt.hour = text[0];
				} else if (opt.hours) {
					if (opt.hours >= 24) {
						opt.hour = 23;
						opt.minute = 59;
						opt.second = 59;
					} else {
						opt.second = 0;
						opt.minute = 0;
						opt.hour = opt.hours;
					}
				} else if (opt.minutes) {
					opt.second = 0;
					opt.minute = opt.minutes % 60;
					opt.hour = (opt.minutes - opt.minute) / 60;
				} else if (opt.seconds) {
					opt.second = opt.seconds % 60;
					opt.minute = ((opt.seconds - opt.second) / 60) % 60;
					opt.hour = ((opt.seconds - opt.second) - (opt.minute * 60)) / 60 / 60;
				}

				var message = methods.checkTime(opt.hour, opt.minute, opt.second);
	
				if (message) {
					$this.html('<div style="color: #F00; font-size: 9px;">Number out of range!</div>');
					$.error(message);
				}

				var hour		= methods.getNumber(opt.hour),
					minute		= methods.getNumber(opt.minute),
					second		= methods.getNumber(opt.second),
					$hour		= $('<div />', { id: 'hour', html: hour, style: 'float: left;' }),
					$minute		= $('<div />', { id: 'minute', html: minute, style: 'float: left;' }),
					$second		= $('<div />', { id: 'second', html: second, style: 'float: left;' }),
					timer		= 0;

				if (opt.displayHours) {
					$this.append($hour);

					if (opt.displayMinutes || opt.displaySeconds) {
						$this.append(separator);
					}
				}

				if (opt.displayMinutes) {
					$this.append($minute);

					if (opt.displaySeconds) {
						$this.append(separator);
					}
				}

				if (opt.displaySeconds) {
					$this.append($second);
				}

				var $separators = $this.children('span');

				methods.checkAlert.call(self, hour, minute, second);

				timer = setInterval(function() {
					if (self.opt.blink) {
						$separators.fadeOut(self.opt.blinkTime, function() {
						    $(this).fadeIn(self.opt.blinkTime);
						});
					}

					if (second == 0) {
						if (minute == 0) {
							if (hour > 0) {
								hour = methods.getNumber(hour - self.opt.decrement);
								minute = 59;
								second = 59;

								$hour.html(hour);
								$minute.html(minute);
								$second.html(second);
							}
						} else {
							minute = methods.getNumber(minute - self.opt.decrement);
							second = 59;

							$minute.html(minute);
							$second.html(second);
						}
					} else {
						second = methods.getNumber(second - self.opt.decrement);
						$second.html(second);
					}

					if (self.opt.finish && second == 0 && minute == 0 && hour == 0) {
						self.opt.finish.call(self);

						clearInterval(timer);
					}

					methods.checkAlert.call(self, hour, minute, second);
				}, 1000);
			});
		}, checkAlert: function(hour, minute, second) {
			var $this	= $(this),
				alert	= this.opt.alert;

			if (alert && $this.css('color') != '') {
				if (hour <= alert.hour && minute <= alert.minute && second <= alert.second) {
					$this.css('color', alert.color);
				}
			}
		}, checkTime: function(hour, minute, second) {
			if (hour < 0 || hour > 24) {
				return 'The hour must be >= 0 or <= 24';
			}

			if (minute < 0 || minute > 59) {
				return 'The minute must be >= 0 or <= 59';
			}

			if (second < 0 || second > 59) {
				return 'The second must be >= 0 or <= 59';
			}
		}, getNumber: function(number) {
			return (number < 10) ? '0' + ((number < 0) ? '0' : number) : number;
		}, set: function(options) {
			return this.each(function() {
				this.opt = $.extend({}, $.fn.chrony.defaults, $(this).data('options'), options);
			});
		}
	};

	$.fn.chrony = function(method) {
		if (methods[method]) {
			return methods[method].apply(this, Array.prototype.slice.call(arguments, 1));
		} else if (typeof method === 'object' || !method) {
			return methods.init.apply(this, arguments);
		} else {
			$.error('Method ' + method + ' does not exist!');
		} 
	};

	$.fn.chrony.defaults = {
		alert			: { color: '#F00', hour: 0, minute: 0, second: 10 },
		blink			: false,
		blinkTime		: 130,
		finish			: undefined,
		decrement		: 1,
		displayHours	: true,
		displayMinutes	: true,
		displaySeconds	: true,
		hour			: 0,
		hours			: undefined,
		minute			: 0,
		minutes			: undefined,
		second			: 0,
		seconds			: undefined,
		text			: undefined
	};

})(jQuery);
